src/v/cluster/tm_stm.cc

```diff
@@ -117,16 +117,77 @@ tm_stm::tm_stm(
       config::shard_local_cfg().tm_violation_recovery_policy.value())
   , _feature_table(feature_table) {}
 
-ss::future<std::optional<tm_transaction>>
-tm_stm::get_tx(const kafka::transactional_id& tx_id) {
+std::optional<tm_transaction>
+tm_stm::do_get_tx(const kafka::transactional_id& tx_id) {
     auto tx = _mem_txes.find(tx_id);
     if (tx != _mem_txes.end()) {
-        co_return tx->second;
+        return tx->second;
     }
     tx = _log_txes.find(tx_id);
     if (tx != _log_txes.end()) {
-        co_return tx->second;
+        return tx->second;
+    }
+    return std::nullopt;
+}
+
+ss::future<std::optional<tm_transaction>>
+tm_stm::get_tx(const kafka::transactional_id& tx_id) {
+    auto tx_opt = do_get_tx(tx_id);
+    if (!tx_opt) {
+        // not found.
+        co_return tx_opt;
+    }
+    auto tx = tx_opt.value();
+    if (
```

Comment:
do we need to exclude this case? it seems that `!tx->transferring` branch will handle it anyway

Reply:
right, nice catch.